doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+time rather than walking the entire `String` and compute the hash code in runtime. For this `String`, the answer
+is already known during the compile time. Therefore, the call to `hashCade()` is replaced with a constant `String`.
+It is powerful because a call with a non-trivial logic is replaced with a simple constant.
+
+Similarly, if the callee compares the `String` that is passed in with some other constant strings, the comparison
+can also be performed during the compile time. The call `equal()` is completely eliminated in the runtime.
+
+Inlining also allows to stack allocate an object better because of a broader scope of things as a result of inlining.
+A return object from the callee to the caller could also be eliminated if the method is inlined. If the caller uses
+the object and then stops using it, it is possible to allocate the object on the stack. When the callee is brought
+into the caller, there is no need to save and restore registers across the call. Register assignment can also be
+done in a better way.
+
+Inlining is an enabler for other optimizations that run later. That is partially the reason why inliner runs
+fairly early in an optimization strategy. Code should be exposed as much as possible and as early as possible
+to other optimizations so that other optimizations can take advantage of what inlining has been done and optimize
+the code better.
+
+# 2. The Downside Of Inlining
+
+The downside is that the method gets much bigger if many callees are inlined into a method. In general, for a
+language like Java that has so many calls and is object oriented, it is always beneficial to inline a method on
+the hot path, unless the callee is really large and it executes lots of instructions. A small method on a hot
+path tends to get inlined in a language like Java because of so much calling and object orientedness built into it.
+
+# 3. Balance Multiple Paths Of Optimizations
+
+There is a division of responsibilities among the paths of optimizations in different parts of the JIT. We do not
+want to pollute every optimization with every other optimization. For example, we do not want to do folding or
+dead tree elimination while we are doing inlining. The optimizations that generate the dead trees could have
+added some code to eliminate the dead trees, but we choose not to do that. Instead, the dead trees will be cleaned
+up by dead tree elimination optimization.
+
+When inlining is taking place, we are aware of the opportunity for other optimizations because of inlining. Such
+as when looking at the arguments of the call at the call site and there is a constant `String`, that should increase
+the priority of inlining the call. It makes it more likely to inline the call because there might be other
+optimizations that can happen downstream because of passing in this constant `String`. But inlining will not do
+that optimization, that optimization will be done separately by other optimization paths when they run.
```

Comment:
You may want to refer to "other optimizations" as "dependent optimizations". This would also allow you to rephrase that last awkward sentence as :  "The inlining pass does not do those dependent optimizations since that logic is already in other optimization passes that will run later."